### PR TITLE
Replace Distribution+TorchDistribution with a thin Distribution mixin

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #732 

This refactors our PyTorch wrapping layer from a *has-a* to an *is-a* relationship, thereby allowing Pyro to use PyTorch-compatible distributions directly in Pyro sample statements.

## Why?

- This brings us much closer to being able to programmatically import all PyTorch distributions.
- This makes it easier to implement new PyTorch-compatible distributions for use in Pyro.

## How?

This PR extends Pyro's `Distribution` class hierarchy with a small mixin class
```
   Distribution
        |
TorchDistributionMixin    torch.distributions.Distribution
                 |           |
               TorchDistribution
```
The mixin allows PyTorch distributions to be subclassed and used directly in Pyro by:
- renaming `.reparameterized`, `.enumerable`, and param names, and
- providing default logic for `.__call__()` and `.score_parts()`
- providing a `.reshape()` property and `Reshape` wrapper class to finely control distribution shapes, e.g.
```diff
- pyro.sample("x", dist.Normal(mu, sigma, extra_event_dims=1), sample_shape=(2,3))
+ pyro.sample("x", dist.Normal(mu, sigma)
+                      .reshape(sample_shape=(2,3), extra_event_dims=1))
```

## Tasks

- [x] Implement a `TorchDistributionMixin` class
- [x] Add a `Reshape` distribution
- [x] Refactor `Normal`
- [x] Refactor `TransformedDistribution`
- [x] Refactor other distributions
- [x] Update all code to use `Reshape` instead of `extra_event_dims`
- [x] Update all code to use `Reshape` instead of `sample_shape`
- [x] Clean up docs

### CODE REVIEW

File: pyro/poutine/trace.py [hunk 1]
```diff
@@ -82,10 +80,7 @@ def log_pdf(self, site_filter=lambda name, site: True):
                     site_log_p = site["log_pdf"]
                 except KeyError:
                     args, kwargs = site["args"], site["kwargs"]
-                    if isinstance(site["fn"], TorchDistribution):
-                        site_log_p = site["fn"].log_prob(site["value"]).sum() * site["scale"]
-                    else:
-                        site_log_p = site["fn"].log_prob(site["value"], *args, **kwargs).sum() * site["scale"]
+                    site_log_p = site["fn"].log_prob(site["value"], *args, **kwargs).sum() * site["scale"]
```
Comment (fritzo) [inline pyro/poutine/trace.py]: @eb8680 Now that `sample_shape` is handled by `Reshape`, we no longer need this `isinstance` kludge.

**Comment by eb8680** [comment]
I like it! Now if we can just automate the `TorchDistribution` wrapper generation then this will be really clean.

> Refactor Histogram to not inherit from `Distribution`

Do we want to allow stochastic functions that aren't `Distribution`s to expose a `log_prob` method?

**Comment by fritzo** [comment]
> Do we want to allow stochastic functions that aren't Distributions to expose a log_prob method?

Yes. This PR reappropriates the name `Distribution` to denote a mixin class that provides Pyro functionality to PyTorch distributions. Due to Python's method resolution order, this pattern prevents us from having a common base class to declare the `.__call__()`+`.log_prob()` interface. In the lack of such a common base class (formerly named `Distribution`), I think the cleanest thing to do is allow duck-typed stochastic functions to have `.log_prob()` methods.

**Comment by fritzo** [comment]
@eb8680 I was unable to eliminate the `Distribution` abstract base class from Pyro due to `isinstance()` logic in lift_poutine.py. Instead I've inserted the `TorchDistributionMixin` class in the existing hierarchy.

File: pyro/distributions/torch/__init__.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
+from __future__ import absolute_import, division, print_function
```
Comment (fritzo) [inline pyro/distributions/torch/__init__.py]: This file is now much closer to being a simple loop over `dir(torch.distributions)`. It still remains to rename parameters and fix `Multinomial` parameter types.

**Comment by fritzo** [comment]
Thanks for reviewing, @eb8680!

**Pull request merged**